rmw_dds_common/src/qos.cpp

```diff
@@ -664,4 +668,47 @@ qos_profile_update_best_available_for_services(const rmw_qos_profile_t & qos_pro
   return result;
 }
 
+rmw_ret_t
+parse_type_hash_from_user_data(
+  const uint8_t * user_data,
+  size_t user_data_size,
+  rosidl_type_hash_t & type_hash_out)
+{
+  RMW_CHECK_ARGUMENT_FOR_NULL(user_data, RMW_RET_INVALID_ARGUMENT);
+  std::vector<uint8_t> udvec(user_data, user_data + user_data_size);
+  auto key_value = rmw::impl::cpp::parse_key_value(udvec);
+  auto typehash_it = key_value.find("typehash");
+  if (typehash_it == key_value.end()) {
+    type_hash_out = rosidl_get_zero_initialized_type_hash();
+    return RMW_RET_OK;
+  }
+  std::string type_hash_str(typehash_it->second.begin(), typehash_it->second.end());
+  if (RCUTILS_RET_OK != rosidl_parse_type_hash_string(type_hash_str.c_str(), &type_hash_out)) {
+    return RMW_RET_ERROR;
+  }
+  return RMW_RET_OK;
+}
+
+rmw_ret_t
+encode_type_hash_for_user_data_qos(
+  const rosidl_type_hash_t & type_hash,
+  std::string & string_out)
+{
+  if (type_hash.version == ROSIDL_TYPE_HASH_VERSION_UNSET) {
+    string_out.clear();
+    return RMW_RET_OK;
+  }
+  auto allocator = rcutils_get_default_allocator();
+  char * type_hash_c_str = nullptr;
+  rcutils_ret_t stringify_ret = rosidl_stringify_type_hash(&type_hash, allocator, &type_hash_c_str);
+  if (RCUTILS_RET_BAD_ALLOC == stringify_ret) {
+    return RMW_RET_BAD_ALLOC;
+  } else if (RCUTILS_RET_OK != stringify_ret) {
+    return RMW_RET_ERROR;
+  }
+  string_out = "typehash=" + std::string(type_hash_c_str) + ";";
+  allocator.deallocate(type_hash_c_str, &allocator.state);
```

Comment:
Since you're in C++ here, you should be careful with allocations like this. For example, if line 709 threw an exception, which is totally possible, then the deallocate would never happen.

Fortunately, since you're in C++ you can use the scope exit pattern to ensure that after allocation has happened successfully, deallocation will absolutely occur, even if there's an exception.

There's a scope exit helper in rclcpp and maybe also in rcpputils? Not sure. In this case you might be able to use a unique_ptr with a custom deleter.

Reply:
Also, since we're potentially being called from a C context, we might need to ensure any exceptions are caught. Since throwing an exception up through a C function is undefined behavior.

Reply:
> There's a scope exit helper in rclcpp and maybe also in rcpputils?

Yeah, it's available in `rcpputils`: https://github.com/ros2/rcpputils/blob/rolling/include/rcpputils/scope_exit.hpp  .  I'd actually prefer using the `RCPPUTILS_SCOPE_EXIT` over a `unique_ptr` since a) `rmw_dds_common` already depends on it, and b) for whatever reason, clang static analysis has an easier time reasoning about `SCOPE_EXIT` than a custom deleter.

Reply:
Done - added scope exit after checking for `BAD_ALLOC` but before anything else.